plugins/modules/gitlab_merge_request_rule.py

```diff
@@ -0,0 +1,565 @@
+                    )
+            final_branches.add(_id)
+        return final_branches
+
+    def _check_protected_branch_exists(self, protected_branch):
+        try:
+            return next(
+                pb.id for pb in self.protected_branches if pb.name == protected_branch
+            )
+        except StopIteration:
+            return None
+
+    def _string_to_set(self, string):
+        try:
+            return set(string.strip().split(",")) if string is not None else None
+        except AttributeError:
+            if isinstance(string, int):
+                return set([string]) if string is not None else None
+            elif isinstance(string, list):
+                return set(string) if string is not None else None
+            return None
+
+    def _set_to_string(self, selfset):
+        return ",".join(str(s) for s in selfset) if selfset is not None else None
+
+    def build_groups(self, wanted_groups):
+        final_groups = set()
+        for grp in wanted_groups:
+            try:
+                final_groups.add(GitlabGroup(self._module, self.gitlab, grp).id)
+            except gitlab.GitlabError as err:
+                self._module.fail_json(
+                    msg="Failed to determine specified group ({})".format(grp), err=err
+                )
+        return final_groups
+
+    def build_users(self):
+        final_users = set()
+        if self.rule_user_ids is not None:
+            if self.rule_usernames is not None:
+                combined_users = set(chain(self.rule_user_ids, self.rule_usernames))
+            else:
+                combined_users = self.rule_user_ids
+        elif self.rule_usernames is not None:
+            combined_users = self.rule_usernames
+        else:
+            return None
+        for usr in combined_users:
+            try:
+                found_user = GitlabUser(self._module, self.gitlab, usr)
+                final_users.add(found_user.id)
+            except gitlab.GitlabError as err:
+                self._module.fail_json(
+                    msg="Failed to determine specified user ({})".format(usr), err=err
+                )
+        return final_users
+
+    def get_project(self, group_name, project_name):
+        return self.gitlab.projects.get("{}/{}".format(group_name, project_name))
+
+    def get_project_group(self, group_name):
+        return self.gitlab.groups.get(group_name)
+
+    def list_approval_rules(self):
+        return self.project.approvalrules.list()
+
+    def check_rule_exists(self, rule_name):
+        try:
+            return True, next(r for r in self.rules if r.name == rule_name)
+        except StopIteration:
+            return False, None
+
+    def _generated_data(self):
+        _name = str(self.rule_name)
+        _approvals_required = int(self.rule_apporvals_required)
+        _protected_branch_ids = (
+            list(self.rule_branches) if self.rule_branches is not None else []
+        )
```

Comment:
This (probably the lines below as well) are overkill: the code already guarantees that `self.rule_branches` is not going to be `None`, in lines 167-171.